hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -271,6 +271,21 @@ object PruneDeadFields {
             // don't memoize right if we are going to elide it during rebuild
             memoizeTableIR(left, requestedType, memo)
         }
+      case TableMultiWayZipJoin(children, fieldName, globalName) =>
+        val gType = requestedType.globalType.fieldOption(globalName)
+          .map(_.typ.asInstanceOf[TArray].elementType)
+          .getOrElse(TStruct()).asInstanceOf[TStruct]
+        val rType = requestedType.rowType.fieldOption(fieldName)
+          .map(_.typ.asInstanceOf[TArray].elementType)
+          .getOrElse(TStruct()).asInstanceOf[TStruct]
+        children.foreach { child =>
+          val dep = child.typ.copy(
+            rowType = TStruct(child.typ.rowType.required, child.typ.rowType.fieldNames.flatMap(f =>
+                child.typ.keyType.fieldOption(f).orElse(rType.fieldOption(f)).map(reqF => f -> reqF.typ)
+              ): _*),
+            globalType = gType)
+          memoizeTableIR(child, dep, memo)
```

Comment:
Can you double check if there should be an extra case for `TableMultiWayZipJoin` in `rebuild` in `PruneDeadFields`?

Reply:
I do not believe it is necessary in this case.

Reply:
I'm thinking there does need to be -- what if I prune some row field in one of the children and not the others? I feel like we need to upcast.

Reply:
(if not equal)

Reply:
I added an upcast rule. Please take a look.